src/devices/bus/epson_qx/multifont.cpp

```diff
@@ -0,0 +1,258 @@
+//-------------------------------------------------
+void multifont_device::device_add_mconfig(machine_config &config)
+{
+	I8039(config, m_mcu, 9_MHz_XTAL);
+	m_mcu->set_addrmap(AS_PROGRAM, &multifont_device::rom_map);
+	m_mcu->prog_out_cb().set(m_i8243, FUNC(i8243_device::prog_w));
+	m_mcu->p1_out_cb().set(FUNC(multifont_device::p1_w));
+	m_mcu->p2_out_cb().set(FUNC(multifont_device::p2_w));
+	m_mcu->bus_out_cb().set(FUNC(multifont_device::write_bus));
+	m_mcu->bus_in_cb().set(FUNC(multifont_device::read_bus));
+	m_mcu->t1_in_cb().set(FUNC(multifont_device::t1_r));
+	m_mcu->t0_in_cb().set(FUNC(multifont_device::t0_r));
+	I8243(config, m_i8243);
+	m_i8243->p4_out_cb().set(FUNC(multifont_device::rom_addr_w<0>));
+	m_i8243->p5_out_cb().set(FUNC(multifont_device::rom_addr_w<4>));
+	m_i8243->p6_out_cb().set(FUNC(multifont_device::rom_addr_w<8>));
+	m_i8243->p7_out_cb().set(FUNC(multifont_device::rom_select_w));
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+void multifont_device::device_start()
+{
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+void multifont_device::device_reset()
+{
+	m_data_out = 0x00;
+	m_data_in = 0x00;
+	m_p2 = 0x00;
+	m_status = 0x00;
+	address_space &space = m_bus->iospace();
+
+	offs_t	iobase = (m_iobase->read() & 0x06) + 0xf8;
+
+	space.install_device(iobase, iobase+1, *this, &multifont_device::map);
+}
+
+uint8_t multifont_device::t0_r()
+{
+	return m_status & 0x01;
+}
+
+uint8_t multifont_device::t1_r()
+{
+	return (m_status & 0x80) >> 7;
+}
+
+void multifont_device::write_bus(uint8_t data)
+{
+	m_status = (m_status & 0x7f) | 0x80;
+	m_data_out = data;
+}
+
+uint8_t multifont_device::read_bus()
+{
+	if (m_p1 & 0x01) {
+		m_status = (m_status & 0xfe);
+		return m_data_in;
+	} else if (m_p1 & 0x02) {
+		uint8_t *rom = m_fonts[m_rom_bank]->base();
+		return rom[m_address];
+	}
+	return 0xff;
+}
+
+void multifont_device::p1_w(uint8_t data)
+{
+	m_p1 = data;
+	if ((m_p1 & 0x80) == 0) {
+		get_slot()->intl_w(ASSERT_LINE);
+	}
+}
+
+void multifont_device::p2_w(uint8_t data)
+{
+	m_p2 = data;
+	m_i8243->p2_w(m_p2 & 0x0f);
+	update_bank_select();
+	m_status = (m_status & 0xfd) | ((m_p2 & 0x20) >> 4);
+}
+
+template<int Shift>
+void multifont_device::rom_addr_w(uint8_t data)
+{
+	m_address = (m_address & ~(0x00f << Shift)) | ((data & 0x0f) << Shift);
+}
+
+void multifont_device::rom_select_w(uint8_t data)
+{
+	m_rom_sel = (data & 0x0f) >> 2;
+	m_address = (m_address & 0xfff) | ((data & 0x03) << 12);
+	update_bank_select();
+}
+
+void multifont_device::update_bank_select()
+{
+	m_rom_bank = ((m_p2 & 0x80) >> 5) | (m_rom_sel & 0x03);
+}
+
+uint8_t multifont_device::read(offs_t offset)
+{
+	if (offset) {
+		get_slot()->intl_w(CLEAR_LINE);
+		return m_status;
+	}
+	m_status = (m_status & 0x7f);
+	return m_data_out;
```

Comment:
You should check for disabled side effects here to avoid issues if a debugger memory view is open.

Reply:
Can you explain what you mean by disabled side effects here? I'm not sure what you want me to be checking for or what issues might arise.

Reply:
If you have a debugger memory view showing an address space, it needs to read from the address space to show what the CPU or whatever other device would see if it read from that address space.  However, you don’t want these reads affecting the state of the system, or it would behave differently when a memory window is open.  In this case, simply having a memory window open showing the address space where the card is mapped will cause the interrupt to be acknowledged if you leave the code like this.

See the discussion on #9215, or look at these read handlers:
* https://github.com/mamedev/mame/pull/9215/files#diff-ac6b09b5779ab8b0dd9ef02cc641a8ed4646bda46bec7769ed6ccebcbf33f388R171
* https://github.com/mamedev/mame/pull/9215/files#diff-ac6b09b5779ab8b0dd9ef02cc641a8ed4646bda46bec7769ed6ccebcbf33f388R268

Or see the documentation for memory accesses from the debugger here: https://docs.mamedev.org/debugger/#memory-accesses

Reply:
(With regards to the documentation, the semantics described there only apply when devices and drivers are well-behaved – there are some badly-behaved drives and devices, but they should be fixed, and we don’t want more of them.)